---
title: "Thursday.rmd"
author: "Elizabeth Garfinkle & Corinne Strawser"
date: "2024-04-29"
output: html_document
---

# Session Description
This session will provide an introduction to cell type annotation and 
differential expression analysis of single-cell RNA-seq data.

# Session Learning Objectives

After completion, you should be able to cell type clusters and perform 
differential expression analysis using scRNA-seq and pseudobulk methods. 
Additionally, you should be able to visualize the output of your differential 
expression analysis. 

# Part I: Introduction to Cell Type Annotation
* Load in integrated COVID PBMC Seurat object from previous day 
* Cell type annotation 
  * Load PBMC cell type reference and visualize data as a UMAP 
  * Find anchors between the PBMC reference data and the query COVID data 
  * Transfer anchors to query COVID data and add cell type information as metadata 
  * Visualize typed clusters and confirm with key cell type markers
* *Activity 1*
  * Tally the number of cells in each type and per condition 
  * Use DimPlot to show cell type clusters for each condition
* Visualize data as a UMAP separated by treatment

# Part II: Introduction to Differential Expression Analysis
* Prepare for differential expression testing
* Differential gene expression between clusters
  * FindAllMarkers
    * Visualize markers on a heatmap
    * Visualize markers on a FeaturePlot
  * FindConservedMarkers
* *Activity 2*
  * Change identity to cell type
  * Visualize cell types as a UMAP split by condition
  * Run FindAllMarkers to identify genes defining each cell type
  * Find cell type markers and visualize as dotplot
* Differential gene expression within cell types between conditions
  * Find genes differentially expressed in a cluster between conditions
  * Explore the results
  * Save DEGs as a .csv file
  * Visualize data as a volcano plot
  * Visualize select genes as a FeaturePlot
  * Visualize select genes as a VlnPlot
  * Finding DEGs in a for-loop
* *Activity 3*
  * Find genes that are differentially expressed between two control and severe patients in CD14 Monocytes
  * Explore results and save as a .csv file
  * Visualize top 3 most significantly different genes using VlnPlot
* Pseudobulked Differential Gene Expression
  * Why pseudobulk?
  * Prepare data for pseudobulk analysis
  * Perform DESeq2 on pseudobulk data
    * Save DEGs as a .csv file
    * Visualize data as a volcano plot
* *Activity 4*
  * Use the same cell type as in Activity 3
  * Find genes that are differentially expressed between two conditions using the pseudobulk method
  * Explore results and save as a .csv file
  * Visualize top 3 most significantly different genes using VlnPlot
  * How do the results from the two DEG analyses compare?
* (Time Permitting) Perform GSEA analysis on a list of DEGs
  * Save GSEA results as a .csv file
  * Visualize data as a bubble plot
* *Session Challenge*
  * Find genes differentially expressed in CD14 Monocytes between severe and mild patients
  * Explore the results and save as a .csv file
  * Visualize top 3 most significantly different genes using a VlnPlot
  * Perform GSEA analysis on the CD14 Monocytes Severe vs. Mild DEGs using the Hallmark geneset
  * Visualize GSEA results as a bubble plot
```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("fgsea")

BiocManager::install("DESeq2")
```
    
```{r setup, include=FALSE}
library(Seurat)
library(SeuratData)
library(ggplot2)
library(tidyverse)
library(stringr)
library(msigdbr)
library(fgsea)
library(data.table)
library(RColorBrewer)
library(ggrepel)
library(DESeq2)

# set seed for reproducibility
set.seed(888)

# set working directory
setwd("./Sessions/beginner_thursday")

# not sure if we will need to do this or not
options(future.globals.maxSize = 8000 * 1024^24)
```

# Part I: Introduction to Cell Type Annotation

## Load in query data

Load in the integrated COVID PBMC Seurat object to cell type 
```{r load_data}
# read in seurat object
data <- readRDS("../../integrated.rds")

# the reference data is SCTranformed so we will use the same approach to normalize the query
data <- SCTransform(data)

# set assay to SCT
DefaultAssay(data) <- "SCT"

# set identity for the dataset -- RPCA integration seurat clusters
Idents(data) <- "rpca_clusters"

# visualize clusters via DimPlot
DimPlot(data, reduction = "umap.rpca", label = T, label.box = T, repel = T)
```

## Load in PBMC reference data 

Load in the PBMC reference Seurat object to serve as the reference for cell typing
This object has three differnet levels of cell type granularity: l1, l2, and l3 -- we will use l1 for this query 
```{r umap}
# read in PBMC reference seurat object
pbmc_ref <- readRDS("pbmc_multimodal_2023.rds")

# set assay to SCT
DefaultAssay(pbmc_ref) <- "SCT"

# set identity to l2 
Idents(pbmc_ref) <- "celltype.l2"

# visualize PBMC cell type clusters via DimPlot
DimPlot(pbmc_ref, label = T, label.box = T, repel = T) #label, label.box, and repel will add labels to the clusters on the DimPlot 
```

## Find anchors and transfer
```{r umap}
# Find anchors between PBMC reference and COVID query 
anchors <- FindTransferAnchors(reference = pbmc_ref, query = data, dims = 1:30,
    reference.reduction = "pca")

# Transfer anchors 
predictions <- TransferData(anchorset = anchors, refdata = pbmc_ref$celltype.l2, dims = 1:30)

# Add query to metadata of COVID object 
data <- AddMetaData(data, metadata = predictions)

# Rename "predicted.id" to "cell_types"
data$cell_types <- data$predicted.id

# Set active identity to new cell type labels and visualize typed clusters via DimPlot
Idents(data) <- data$cell_types
DimPlot(data, reduction = "umap.rpca", label = T, label.box = T, repel = T)

# Tally number of cells in each cell type 
table(data$cell_types)
```

## Verify cell typing with Feature Plots for key cell type markers
```{r}
# Verify T regs
# CD3+ FOXP3+
FeaturePlot(data, features = c("CD3E", "FOXP3"))
```

## Save your final typed COVID Seurat object
```{r}
saveRDS(data, "data_typed.rds")
```

## Additional resources 
Curated single cell databases: 
https://cellxgene.cziscience.com 
https://azimuth.hubmapconsortium.org
https://scpca.alexslemonade.org 

SingleR package: 
https://github.com/dviraran/SingleR

## Activity 1:
### Tally the number of cells in each type and per condition (hint: use the "group_per_sample" metadata)
```{r}


```

### Use DimPlot to show cell type clusters for each condition  
```{r}


```








  
# Part II: Introduction to Differential Expression Analysis 

## Let's set up for DEG analysis
```{r setUp}
# First, make sure that our data layers are joined
data@assays$RNA@layers %>% 
  names()
# if it's not joined you'll have several counts and data layers

# join layers after integration if not 
# data[["RNA"]] <- 
#   JoinLayers(data[["RNA"]])

# check / set default assay
DefaultAssay(data) <- 
  "RNA"

# Second, check to make sure relevant meta data is in object
data@meta.data %>%
  head()

# if not, add it in
# read in meta data frame
# Idents(merged) <- colnames(merged)
# meta <- read.table("240429_final_meta.csv",
#                    sep = ",")
# merged <- AddMetaData(merged,
#                       meta)

# set identity for the dataset
Idents(data) <- 
  "rpca_clusters"

# how many cells do we have per condition?
table(data$rpca_clusters, data$group_per_sample)

# how many cells do we have per patient?
table(data$rpca_clusters, data$sampleID)
```

## Visualize data as a UMAP

Set identity for Seurat object as "group_per_sample", which is the slot in our
metadata where the COVID conditions are stored
```{r umap}
# plot UMAP separated by treatment
DimPlot(object = data,
        group.by = "group_per_sample",
        reduction = "umap.rpca") &
  xlab("UMAP1") &
  ylab("UMAP2")
```

# Differential Gene Expression

In this section we will learn how to perform differential expression analysis
using Seurat built-in functions. We will use the RNA assay.
```{r}
# check to see if RNA data has been normalized yet
GetAssayData(object = data,
             assay = "RNA",
             layer = "data")[1:10,1:10]
# integers typically indicate data has not been normalized
# floats typically indicate data has been normalized
```

## Between clusters

We can use any slot in the metadata for performing DE testing. We will start
with clusters.

### Find markers that define clusters with FindAllMarkers

First, we will confirm that our identity is set to our rpca_cluster metadata slot.
We will then perform DE testing for all clusters in our data using the
FindAllMarkers function which will identify genes that are differentially
expressed in each identity group by comparing it to all of the others. Note that
this method does not identify markers that are specific to a single group, and 
thus, closely-related groups may have overlapping markers.

```{r FindAllMarkers}
# make sure identity is set to rpca clusters
Idents(data) <- 
  "rpca_clusters"
levels(data) %>%
  unique()

# look at umap
DimPlot(object = data,
        reduction = "umap.rpca",
        label = T,
        label.box = T,
        repel = T) &
  xlab("UMAP1") &
  ylab("UMAP2")

# find markers for each cluster
markers <- 
  FindAllMarkers(object = data,
                 min.pct = 0.25,
                 only.pos = T)
# this finds genes that differentially expressed for each cluster selecting
# for genes that are expressed in >25% of cells and only considering genes
# that have a positive log2FC

# reduce data
top_markers <- 
  markers %>%
  as.data.frame() %>%
  group_by(cluster) %>% # group by cluster
  slice_min(p_val_adj, 
            n = 3,
            with_ties = F) # sort by adjusted p-value and get bottom 3

# explore top markers
head(top_markers)

# save markers as a dataframe
write.csv(x = top_markers, 
          file = "FindAllMarkers_top3_cluster_markers.csv")
```

### Visualize markers as a heatmap

Sometimes it's nice to see globally which markers are defining specific groups 
and a heatmap is a great way to do that. We will want to use normalized counts
for this so that each group has comparable values. This will show the normalized
expression for each gene for each cell in each identity.

```{r}
# get list of top marker genes; including each gene only once
genes <- 
  top_markers$gene %>%
  unique()

# change to RNA assay
DefaultAssay(data) <- 
  "RNA"

# scale data for visualization
data <-
  ScaleData(object = data,
            assay = "RNA")

# make heatmap
small <-
  subset(x = data,
         downsample = 50)
DoHeatmap(object = small,
          features = genes,
          group.by = "rpca_clusters")
ggsave(filename = "top3_cluster_markers.jpeg",
       width = 8,
       height = 8)
```

### Visualize markers as a FeaturePlot

```{r}
# plot some top gene per cluster
top_markers <- 
  markers %>%
  as.data.frame() %>%
  group_by(cluster) %>% # group by cluster
  slice_min(p_val_adj, 
            n = 1,
            with_ties = F) # sort by adjusted p-value and get bottom 1
genes <- 
  top_markers$gene

# make FeaturePlot
FeaturePlot(object = data,
            reduction = "umap.rpca",
            features = genes,
            ncol = 5,
            label = T) &
  xlab("UMAP1") &
  ylab("UMAP2")
```

### Find markers that define clusters with FindConservedMarkers

We can use FindConservedMarkers to identify genes that are conserved between the
conditions for each identity class and are differentially expressed compared to 
other identity classes. This is helpful to identify true marker genes for each
cluster. Often the results of FindConservedMarkers will be similar to FindAllMarkers.
We won't run it here but it's something to be aware of.

```{r FindConservedMarkers, eval = FALSE}
conserved <- 
  FindConservedMarkers(data,
                       ident.1 = "0",
                       grouping.var = "group_per_sample")
```

## Activity 2

### Change the identity class to *cell type* and make a UMAP colored by cell type and split by COVID condition
```{r}



```

### Use FindAllMarkers to identify genes defining each cell type
```{r}



```

### Make a DotPlot for the top 3 markers per cell type (hint: how do you get information for how to run a function?)

Why might a dot plot be useful for showing scRNA-seq data?
```{r}



```


## Within clusters between conditions

If we have two conditions, we likely want to compare them! To do that, we can 
use the FindMarkers function.

### Find genes differentially expressed in one cluster between conditions

Here we will find the genes that are differentially expressed between a cluster
shared between control and severe COVID patients. We want genes that are expressed 
in at least 10% of the cells in each identity and have a log2FC cut-off as 0.25 (1.2X).
```{r}
# how many cells do we have per condition?
table(data$rpca_clusters, data$group_per_sample)

# change identity class
Idents(data) <- 
  "rpca_clusters"

# set assay
DefaultAssay(data) <- 
  "RNA"

# Find DEGs for cluster 0
degs <- 
  FindMarkers(object = data,
              ident.1 = "severe",
              ident.2 = "control",
              group.by = "group_per_sample",
              subset.ident = "0",
              min.pct = 0.1,
              logfc.threshold = 0.25)
```

### Explore the results
```{r}
# look at first 20 differentially expressed genes
degs %>%
  head(n = 20)

# how many significantly different genes are there?
degs[degs$p_val_adj <= 0.05,] %>%
  nrow()
```

The results data frame has the following columns :

* p_val : p_val (unadjusted)
* avg_log2FC : log fold-change of the average expression between the two groups. 
** Note : positive values indicate that the feature is more highly expressed in the first group.
* pct.1 : percentage of cells where the feature is detected in the first group
* pct.2 : percentage of cells where the feature is detected in the second group
* p_val_adj : adjusted p-value, based on Bonferroni correction using all features in the dataset.

### Save DEGs as a .csv file

We usually want to export the files to share with others and saving them as 
a .csv file is a great way to do that.

```{r}
write.csv(x = degs, 
          file = "cluster_0_severe_vs_control_degs.csv")
```

### Visualize data as a volcano plot

A volcano plot allows us to see gene expression changes and their significance 
for a lot of genes at once.

```{r}
# log transform the padj values for visualization
degs$log.padj <- 
  -log10(degs$p_val_adj)

# label whether the genes were significant
degs$significant <- 
  ifelse(degs$p_val_adj <= 0.05,
         "p < 0.05",
         "p > 0.05")

# pull out labels for the top 20 most differentially expressed genes
labels <- 
  degs %>%
  arrange(desc(abs(avg_log2FC))) %>%
  slice_head(n = 20)

# make volcano plot
volcano <- 
  ggplot(data = degs, 
         aes(x = avg_log2FC, 
             y = log.padj,
             color = significant)) +
  geom_point() +
  scale_color_manual(values = c("darkmagenta", "grey")) +
  geom_label_repel(data = labels,
                   aes(label = rownames(labels)),
                   max.overlaps = 20,
                   show.legend = FALSE) +
  ggtitle("Cluster 0 Differentially Expressed Genes\nSevere COVID19 vs. Control") +
  ylab("-log(padj)") +
  xlab("log2FC") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5),
        text = element_text(size = 12))
volcano
```

### Visualize select genes as a FeaturePlot

A feature plot can allow us to map gene expression onto our UMAP. We can also 
split it based on a slot in the metadata, such as treatment.

```{r}
# make feature plot of one down-regulated and two up-regulated genes from volcano plot
FeaturePlot(object = data,
            reduction = "umap.rpca",
            features = c("SOCS3", "TCEB2"),
            split.by = "group_per_sample",
            label = T,
            order = TRUE) & # pulls positive cells to the front 
  xlab("UMAP1") &
  ylab("UMAP2")
```

### Visualize select genes as a VlnPlot

A violin plot allows us to see gene expression levels within each identity.

```{r}
# make violin plot
VlnPlot(object = data,
        features = c("SOCS3"),
        split.by = "group_per_sample",
        group.by = "rpca_clusters",
        cols = c("#0072B2","#CC79A7","#009E73"),
        ncol = 1,
        pt.size = 0) # removes the black dots for each cell
```

### (Time Permitting) Intro to for-loops in R

In your dataset you'll likely have a few if not a dozen or more clusters and/or
cell types. Thus, instead of writing out code to do the same thing over and over
(like find DEGs) it is useful and more reproducible to iterate over elements in
a list, dataframe, vector, matrix, etc. The for loop can be used to execute a
group of statements or functions repeatedly depending on the number of elements
in your object. 

```{r}
# Example 1
fruits <- c("apple", "banana", "cherry")

for (x in fruits) {
  print(x)
}

# Example 2
my_fav_nums <- c(4,5,6)

for (x in my_fav_nums){
  y <- x + 1
  print(y)
}
```

We could also find DEGs for several clusters at a time using a for-loop

```{r}
# select only clusters that have >3 cells in both groups
df <- 
  table(data$rpca_clusters, 
        data$group_per_sample) %>%
  as.data.frame() %>%
  pivot_wider(names_from = Var2,
              values_from = Freq) %>%
  as.data.frame()
clusters <-
  df[df$control >= 3 & df$severe >= 3, "Var1"] %>%
  as.character()

# Find DEGs for each cluster
for (cell in clusters){
    # Find the markers
    markers <- 
      FindMarkers(object = data,
                  ident.1 = "severe",
                  ident.2 = "control",
                  group.by = "group_per_sample",
                  subset.ident = cell,
                  min.pct = 0.1,
                  logfc.threshold = 0.25)
  
    # make a variable for a filename
    filename <- 
      paste0("cluster_",cell, "_severe_vs_control_degs.csv")
    
    # save our DEGs
    write.csv(markers, 
              file=filename)
}
```

## Activity 3

### Find differentially expressed genes for CD14 Monocytes using scRNA-seq methods
```{r}



```

### Save as a .csv file
```{r}



```

### Visualize as a VlnPlot
```{r}



```


# Pseudobulked Differential Gene Expression

https://satijalab.org/seurat/articles/de_vignette

## Why pseudobulk?
The DE analysis in Seurat treats each cell as an independent sample often 
resulting in inflated p-values as single cells isolated from the same biological
sample are not *truly* independent of one another. Thus, the methods used to
perform DE analysis in Seurat identify genes that vary across individual cells 
not necessarily across a population. To infer DE genes across a population between
conditions we need to pseudobulk our cells.

## Important note:
In order to perform pseudobulked DEG analysis you must have at least 3
biological replicates.

## Prepare data for pseudobulk analysis
First, generate DEG tables from pseudobulk clusters, treating all cells
from each ident from each sample as independent groups of cells/data.
Then, graph each of those DEG datasets using a volcano plot.

```{r}
# what do our cells look like before pseudobulking?
Cells(data) %>%
  head()

# create pseudobulked data for clusters across condition
pb_data <- 
  AggregateExpression(object = data, 
                      assays = "RNA", 
                      return.seurat = T, 
                      group.by = c("group_per_sample", 
                                   "donor", 
                                   "rpca_clusters"))

# what do our cells look like after pseudobulking?
Cells(pb_data) %>%
  head()
# condition_donor_cluster

# add a column to the metadata that is cluster_condition
pb_data$cluster_condition <- paste(pb_data$rpca_clusters,
                                   pb_data$group_per_sample,
                                   sep = "_")
pb_data$cluster_condition %>%
  head()

# how many samples do we have at each level?
pb_data$cluster_condition %>%
  table()

# set default assay to RNA
DefaultAssay(pb_data) <- 
  "RNA"

# perform DE testing on pseudbulked data for same cluster as above
Idents(pb_data) <- 
  "cluster_condition"
pseudobulked_degs <- 
  FindMarkers(object = pb_data, 
              ident.1 = "0_severe",
              ident.2 = "0_control",
              test.use = "DESeq2")
pseudobulked_degs %>%
  head(n = 10)

# how many significant DEGs do we have?
pseudobulked_degs[pseudobulked_degs$p_val_adj <= 0.05,] %>%
  nrow()
```

### Visualize data as volcano plots
```{r}
# log transform the padj values for visualization
pseudobulked_degs$log.padj <- 
  -log10(pseudobulked_degs$p_val_adj)

# label whether the genes were significant
pseudobulked_degs$significant <- 
  ifelse(pseudobulked_degs$p_val_adj <= 0.05,
         "p < 0.05",
         "p > 0.05")

# pull out labels for the top 20 most differentially expressed genes
labels <- 
  pseudobulked_degs %>%
  arrange(desc(abs(avg_log2FC))) %>%
  slice_head(n = 20)

# make volcano plot
volcano <- 
  ggplot(data = pseudobulked_degs, 
         aes(x = avg_log2FC, 
             y = log.padj,
             color = significant)) +
  geom_point() +
  scale_color_manual(values = c("darkmagenta", "grey")) +
  geom_label_repel(data = labels,
                   aes(label = rownames(labels)),
                   max.overlaps = 20,
                   show.legend = FALSE) +
  ggtitle("Cluster 0 Differentially Expressed Genes\nSevere COVID19 vs. Control") +
  ylab("-log(padj)") +
  xlab("log2FC") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5),
        text = element_text(size = 12))
volcano
```

## Activity 4:

### Create a pseudobulk dataset for *cell types*
```{r}



```

### Find differentially expressed genes for CD14 Monocytes using the pseudobulk method
```{r}



```

### How many FEWER DEGs do we have with the pseudobulk method? Compare a gene that is changed in the same direction in both and a gene that is only in the pseudobulk method visually. How do they compare?
```{r}



```


## Perform GSEA analysis on the cluster 0 differentially expressed genes

Gene set enrichment analysis (GSEA) helps us to see if several genes in a
particular process or pathway are changing together. We will use the
Gene Ontology Molecular Functions gene set from MSIGDB and the fgsea package.
There are several GSEA packages and gene sets you can use. You can even create 
your own gene sets!

```{r}
Idents(data) <- "rpca_clusters"
degs <- 
  FindMarkers(object = data,
              ident.1 = "severe",
              ident.2 = "control",
              group.by = "group_per_sample",
              subset.ident = "0",
              min.pct = 0.1,
              logfc.threshold = 0.25)

# look at genesets available
View(msigdbr_collections())

# select dataset and then aggregate the genes into a list
# each element of the list will be a named character vector
# the name of the vector is the gene set name ("gs_name")
# the elements of the vector are genes in the gene set ("gene_symbol")

go_mf <- 
  msigdbr(species = "Homo sapiens",
          category = "C5",
          subcategory = "GO:MF") %>% 
  split(x = .$gene_symbol,
        f = .$gs_name)
View(go_mf)

# arrange DEGs for GSEA 
degs$gene <- 
  row.names(degs)
deg <- 
  degs %>%
  arrange(desc(avg_log2FC)) %>% 
  select(gene, avg_log2FC)

# convert to named list 
ranks <- 
  deframe(deg) 
head(ranks)

# perform GSEA
fgsea_res <- 
  fgseaMultilevel(go_mf,
                  ranks,
                  minSize = 15, # min size of gene set to test
                  maxSize = 500, # max size of a gene set to test
                  nPerm = 1000) %>% # number of permutations to test
  arrange(padj)

# take a look at results
head(fgsea_res)

# save results as a .csv file (use fwrite to handle leadingEdge list)
fwrite(fgsea_res, "cluster0_severe_vs_control_GOMF_GSEA.csv")
```

The GSEA results dataframe has the following columns:

* pathway -- name of the pathway
* pval -- an enrichment p-value
* padj -- a BH-adjusted p-value
* log2err -- the expected error for the standard deviation of the P-value
* ES -- enrichment score, same as in Broad GSEA implementation
* NES -- ES normalized to mean enrichment of random samples of the same size
* size -- size of the pathway after removing genes not present
* leadingEdge -- vector with indexes of leading edge genes that drive the 
                    enrichment

### Visualize data as a bubble plot

It's always nice to display our data as a graph! We will incorporate information
about the pathways enriched, the adjusted p-value, the size of the pathway, and
the normalized enrichment score into a bubble plot.

```{r}
# get only the top 10 pathways
fgsea_plot <- 
  fgsea_res[1:10,]
# make bubble plot
ggplot(fgsea_plot,
       aes(x = NES,
           y = pathway,
           size = size,
           color = padj)) +
  geom_point() +
  scale_color_gradient2(high = "#f517d3",
                        mid = "#ffffff",
                        low = "#0513d8",
                        midpoint = 0.05)
```

## Session challenge:

### Find genes differentially expressed in CD14 Monocytes between severe and mild patients
```{r}



```

### Explore the results and save as a .csv file
```{r}



```

### Visualize top 3 most significantly different genes using a VlnPlot
```{r}



```

### Visualize top 3 most significantly different genes using VlnPlot
```{r}



```

## Perform GSEA analysis on the CD14 Monocytes Severe vs. Mild DEGs using the Hallmark geneset
```{r}



```

### Visualize GSEA results as a bubble plot
```{r}



```
